File: src/lib/structure.ts
import { filterPresentRows } from "./helpers.ts"
import { Block, LayoutError, Structure, StructureCell } from "./types.ts"

export function parseStructure(block: Block): Structure {
  const { lines, indexToLineNr } = filterPresentRows(block.lines, block.startLineNr)

  const s: Structure = { rows: [], separators: [], rowToLineNr: indexToLineNr }

  lines.forEach((line, index) => {
    if (!isStructureLineValid(line)) {
      throw new LayoutError(`Invalid structure row ${index + 1}: '${line}'`, indexToLineNr[index])
    }
  })
  const columns = countColumns(lines)

  for (const line of lines) {
    const row: StructureCell[] = []
    for (let i = 0; i < columns; i++) {
      const value = line.substring(i * 3, i * 3 + 3).trim()
      if (value.length === 0) {
        row.push(null)
      } else if (value === "||") {
        row.push("separator")
      } else {
        row.push({ keyIndex: Number(value) })
      }
    }

    s.rows.push(row)
  }

  // store separator indexes
  s.rows[0].forEach((cell, cellIndex) => {
    if (cell === "separator") {
      s.separators.push(cellIndex)
    }
  })

  // Make sure every row has the same separators
  s.rows.forEach((row, rowIndex) => {
    for (const separator of s.separators) {
      if (row[separator] !== "separator") {
        throw new LayoutError(`Mismatching structure separators at row #${rowIndex + 1}`)
      }
    }
  })

  return s
}

const STRUCTURE_RE = /^([\d ][\d ] |\|\| )*( \d|\d\d?)$/

export function isStructureLineValid(line: string): boolean {
  return STRUCTURE_RE.test(line.trimEnd())
}

export function countColumns(validLines: string[]): number {
  const maxLength = validLines.reduce((max, line) => line.length > max ? line.length : max, 0)
  return Math.ceil(maxLength / 3)
}

File: internal/parser/structure.go
package parser

import (
	"math"
	"regexp"
	"strconv"
	"strings"

	"github.com/elentok/qmkmd/internal/layout"
)

var structureRE = regexp.MustCompile(`^([\d ][\d ] |\|\| )*( \d|\d\d?)$`)

func ParseStructure(block layout.Block) (layout.Structure, error) {
	present := layout.FilterPresentRows(block)
	s := layout.Structure{
		Rows:       [][]layout.StructureCell{},
		Separators: []int{},
		RowToLine:  present.IndexToLineNr,
	}

	for i, line := range present.Lines {
		if !IsStructureLineValid(line) {
			return layout.Structure{}, &layout.LayoutError{
				Msg:    "Invalid structure row " + strconv.Itoa(i+1) + ": '" + line + "'",
				LineNr: present.IndexToLineNr[i],
			}
		}
	}

	columns := CountColumns(present.Lines)
	for _, line := range present.Lines {
		row := make([]layout.StructureCell, 0, columns)
		for i := 0; i < columns; i++ {
			start := i * 3
			end := start + 3
			if start > len(line) {
				row = append(row, layout.StructureCell{})
				continue
			}
			if end > len(line) {
				end = len(line)
			}
			value := strings.TrimSpace(line[start:end])
			switch {
			case value == "":
				row = append(row, layout.StructureCell{})
			case value == "||":
				row = append(row, layout.StructureCell{Kind: layout.CellSeparator, IsOccupied: true})
			default:
				keyIndex, err := strconv.Atoi(value)
				if err != nil {
					return layout.Structure{}, err
				}
				row = append(row, layout.StructureCell{KeyIndex: keyIndex, Kind: layout.CellKey, IsOccupied: true})
			}
		}
		s.Rows = append(s.Rows, row)
	}

	if len(s.Rows) == 0 {
		return s, nil
	}

	for cellIndex, cell := range s.Rows[0] {
		if cell.Kind == layout.CellSeparator {
			s.Separators = append(s.Separators, cellIndex)
		}
	}

	for rowIndex, row := range s.Rows {
		for _, separator := range s.Separators {
			if separator >= len(row) || row[separator].Kind != layout.CellSeparator {
				return layout.Structure{}, &layout.LayoutError{
					Msg: "Mismatching structure separators at row #" + strconv.Itoa(rowIndex+1),
				}
			}
		}
	}

	return s, nil
}

func IsStructureLineValid(line string) bool {
	return structureRE.MatchString(strings.TrimRight(line, " "))
}

func CountColumns(validLines []string) int {
	maxLength := 0
	for _, line := range validLines {
		if len(line) > maxLength {
			maxLength = len(line)
		}
	}
	return int(math.Ceil(float64(maxLength) / 3.0))
}
